Update the doc about the offline update (#669)
foundriesio/docs: source/user-guide/offline-update/offline-update.rst

* ``--wave`` - indicates that this is an update for a production device and ``<target-name>`` refers to *Wave Target* (see the note below)
* ``<days>`` - Offline artifact validity period in days

.. note::
    Use ``fioctl waves init`` command to generate :ref:`Wave Target <ref-rm-wave>`.

    Use ``fioctl waves complete`` command to turn *Wave Target* into :ref:`Production Target <ref-production-targets>`.

    Ensure that the target device is a *Production* device, see :ref:`Manufacturing Process for Device Registration <ref-factory-registration-ref>` for more details.

.. note::
    In order to download all artifacts, ``fioctl`` requires token with scopes: ``targets:read``, ``ci:read``.

Controlling the Expiration Time of the Offline Update Bundle
------------------------------------------------------------

The bundle obtained through the ``fioctl targets offline-update`` command comes with an expiration time.
If the expiration time of the bundle has passed, the offline update will fail.

.. sidebar:: Error if root or targets metadata expires before the specified expiration time

>     .. code-block:: bash

Reviewer: This doesn't look too good on the published page, can we move this to a standard note/code block?
![Screenshot from 2024-02-20 09-16-07](https://github.com/foundriesio/docs/assets/21198569/c182a1ff-e7d7-476a-acad-6148a3e55d05)

Reviewer: Maybe something like:

Use the --expires-in-days parameter of the fioctl targets offline-update command to set the desired expiration time of the bundle. If the bundle installation fails with the error below:

block

It means that TUF root or targets metadata expires sooner than the date specified in the parameter. To fix this problem, either decrease the parameter value or refresh the root/targets metadata accordingly.

Author: ok

        = Targets metadata expire (2025-01-28T16:38:23Z) before the specified expiration time (2025-02-06T09:27:35Z)

Use the ``--expires-in-days`` parameter of the ``fioctl targets offline-update`` command to set the desired expiration time of the bundle.
If the command fails because the root or targets metadata expires sooner than the date specified in the parameter, then
either decrease the parameter value or refresh the root/targets metadata accordingly and then re-run the command.

To refresh root metadata you should :ref:`rotate TUF root role key <ref-offline-tuf-root-key-rotation>`.
The expiration time is set to one year since the moment of the latest root key rotation.

To refresh targets role metadata use one of the following depending on targets type, CI or wave/production.

* CI targets — Trigger a new CI build, it will create a new target and update CI targets role metadata expiration time to 1 year since the moment of creation.
* Wave or production targets — Create a new wave for the given target version.
  Use ``--expires-days`` or ``--expires-at`` parameters of the ``fioctl waves init`` command to set a desired expiration time.

> Therefore, the ``--expires-in-days`` parameter in the primary knob to tune the bundle's expiration time up to 1 year (the maximum validity period of root metadata).

Reviewer: Primary knob is not clear here, I'd prefer explicitly stating the command.

Author: This is all about the same command `fioctl targets offline-update`. That command has many parameters, and the ``--expires-in-days`` is just one of them.

If I put something like `fioctl targets offline-update --expires-in-days` it won't be correct, because the command requires many other parameters which are specified above.

Maybe, there is some other option?

Maybe, `Therefore, the ``--expires-in-days`` parameter of the command` or `Therefore, the ``--expires-in-days`` parameter of the bundle download command`?

Understanding the Math Behind the Offline Update Bundle Expiration Time
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The expiration time of the bundle is determined by the expiration times of the TUF metadata it encompasses.
Specifically, it equals the minimum value among the expiration times across all TUF roles' metadata.

* CI/Wave/Production *root* role metadata
    The expiration time is set to 1 year from the moment when of the latest root key was added or rotated.
    The other commands that modify the TUF root metadata do not extend its expiration.
    It is possible to set the TUF root expiration time to any value through the API.

* CI *timestamp*, *snapshot*, and *targets* roles metadata
    The default expiration time is set to 1 year since the last successful CI build.
    If there are no builds for a year, the expiration is automatically extended by one month every month.
    A user can overwrite the default value using the factory config parameter :ref:`tuf.targets_expire_after <def-tuf-expiration>`.

* Wave/Production *timestamp* roles metadata
    The expiration time is set to 7 days.

>     The `TUF specification`_ recommends setting a short expiration date and re-signing it frequently.

Author: It is not "production artifacts" this is about wave & production timestamp metadata, and the section is called so "Wave/Production *timestamp* roles metadata".

Maybe "short expiration date for the metadata"?

>     The `TUF specification`_ recommends setting a short expiration date and re-signing it frequently.

Reviewer: "short expiration date for the metadata" is better

    This allows clients to quickly detect if they are being prevented from obtaining the most recent metadata ("indefinite freeze attacks").

>     The FoundriesFactory automatically refreshes the metadata for an additional 7 days just before expiration.

Reviewer: This is not clear for me. Why and how does this work?

Author: "Why?" - `The `TUF specification`_ recommends setting a short expiration date and re-signing it frequently`

"how does this work" - this is an internal stuff of the backend implementation, not sure we need to explain low-level technical details here.

* Wave/Production *snapshot* and *targets* roles metadata
    The default expiration time is set to 1 year.

>     A user can overwrite the default value using the ``--expires-days`` or ``--expires-at parameter`` of the ``fioctl wave init`` command.

Reviewer: ```suggestion
    A user can overwrite the default value using the ``--expires-days`` or ``--expires-at`` parameters of the ``fioctl wave init`` command.
```

Author: I guess it should be a singular form, but out of the `` scope indeed.

comes with an expiration time. If the expiration time of the bundle has passed, the offline update will fail.

.. sidebar:: Error if root or targets metadata expires before the specified expiration time

    .. code-block:: bash

        Getting CI Target details; target: intel-corei7-64-lmp-2377, tag: master...
        Refreshing and downloading TUF metadata for Target intel-corei7-64-lmp-2377 to 2377/tuf...
        ERROR: Failed to download TUF metadata: HTTP error during POST 'https://api.foundries.io/ota/factories/<factory>/targets/intel-corei7-64-lmp-2377/meta/': 400 BAD REQUEST
        = Root metadata expire (2024-07-06T07:56:57Z) before the specified expiration time (2025-02-11T09:17:39Z)

        Getting production Target details; target: intel-corei7-64-lmp-2356, tag: master...
        Refreshing and downloading TUF metadata for Target intel-corei7-64-lmp-2356 to 2356/tuf...
        ERROR: Failed to download TUF metadata: HTTP error during POST 'https://api.foundries.io/ota/factories/<factory>/targets/intel-corei7-64-lmp-2356/meta/': 400 BAD REQUEST
        = Targets metadata expire (2025-01-28T16:38:23Z) before the specified expiration time (2025-02-06T09:27:35Z)

> Use the ``--expires-days`` parameter of the ``fioctl targets offline-update`` command to set the desired expiration time of the bundle.

Reviewer: has the command changed or is it just a mistake?
```suggestion
Use the ``--expires-in-days`` parameter of the ``fioctl targets offline-update`` command to set the desired expiration time of the bundle.
```

Author: That's mistake, good catch, thx.

Author: done

If the command fails because the root or targets metadata expires sooner than the date specified in the parameter,
refresh the root or targets accordingly and then re-run the command.

To refresh root metadata you should :ref:`rotate TUF root role key <ref-offline-tuf-root-key-rotation>`.
The expiration time is set to one year since the moment of the latest root key rotation.

To refresh targets role metadata use one of the following depending on targets type, CI or wave/production.

* CI targets - Trigger a new CI build, it will create a new target and update CI targets role metadata expiration time to 1 year since the moment of creation.
* Wave or production targets - Create a new wave for the given target version.
  Use ``--expires-days`` or ``--expires-at`` parameters of the ``fioctl waves init`` command to set a desired expiration time.

> Therefore, the ``--expires-days`` parameter in the primary knob to tune the bundle's expiration time up to 1 year (the maximum validity period of root metadata).
> Effectively, this parameter sets the expiration time for the bundle's copy of the TUF timestamp role metadata, and does not affect the factory's metadata.
> 
> Root and/or CI/wave/production targets refreshing serves as the secondary mechanism that should be applied
> if the desired expiration time occurs later than the root's and/or the targets' expiration, respectively.

Reviewer: I think explicitly setting the command makes this clearer. Also for "wave/production" we've been using production targets across our docs if I'm not mistaken.

```suggestion
Therefore, the ``--expires-in-days`` parameter in the `fioctl targets offline-update` command tunes the bundle's expiration time up to 1 year (the maximum validity period of root metadata).
Effectively, this parameter sets the expiration time for the bundle's copy of the TUF timestamp role metadata, and does not affect the factory's metadata.

Root and/or CI/production targets refreshing serves as the secondary mechanism that should be applied
if the desired expiration time occurs later than the root's and/or the targets' expiration, respectively.
```

Author: > Also for "wave/production" we've been using production targets across our docs if I'm not mistaken.

There are wave and production targets, they are different. The bundle downloading command allows to download one these targets. The "mechanism" to refresh them is the same though.

Author: > I think explicitly setting the command makes this clearer.

@vanmaegima What exactly do you mean?

.. note::
    In order to download all artifacts, ``fioctl`` requires token with scopes: ``targets:read``, ``ci:read``.

Controlling the Expiration Time of the Offline Update Bundle
------------------------------------------------------------

The offline update content, referred to as the "bundle," obtained through the ``fioctl targets offline-update`` command,

> comes with an expiration time. If the expiration time of the bundle has passed, the offline update will fail.

Author: > Just IMHO this paragraph mixes too many things together.

Maybe. If so this is a consequence of mixing many "things" in the command that generates and downloads content for offline update.

The paragraph's aim is to provide a user with a guide/instruction on how to get the offline update bundle with a desired expiration time.

I'll let @caiotpereira and @vanmaegima and @Tim-Anderson to judge about the paragraph style since they know better what customers need.

> comes with an expiration time. If the expiration time of the bundle has passed, the offline update will fail.

Reviewer 2: By "mixing" things I mean that a sentence which is describing the "expiration time" also tries to define a term "bundle".
These two things would look more organic if they are defined separately.

It is really hard to comprehend a sentence which introduces two new terms at once in a rather convoluted way.

> comes with an expiration time. If the expiration time of the bundle has passed, the offline update will fail.

Author: There is no definition of the new term "bundle", this is just alias/another name was the term introduced earlier.

> comes with an expiration time. If the expiration time of the bundle has passed, the offline update will fail.

Reviewer 2: This statement `The offline update content, referred to as the "bundle," obtained through the ``fioctl targets offline-update`` command, ...` is a definition... eghmmm... by definition.

I you prefer to not define it separately, then I suggest to simply remove the `referred to as the "bundle` part of it from the sentence, and use `offline update content` everywhere. That sentence is hard to read for me; so it must be harder for an ordinary person.

I know it for sure, trust me, as I'm also a master to write long complex sentences.

> comes with an expiration time. If the expiration time of the bundle has passed, the offline update will fail.

Author: I can make similar statement.

It is easier to read the sentence with the "bundle" for me; so it must be the same for majority of other people. I know it for sure, trust me, I am a master of writing simple long sentences :).
I am kidding of course.

Again, this is very subjective. IMO, the bundle simplifies the further content since using `offline update content` instead makes the text more complex. For example, "bundle's TUF metadata" vs "TUF metadata of offline update content".

I recon this is up to the customer support team to decide on such subjective stuff since they deal with customers closely.

        Refreshing and downloading TUF metadata for Target intel-corei7-64-lmp-2377 to 2377/tuf...
        ERROR: Failed to download TUF metadata: HTTP error during POST 'https://api.foundries.io/ota/factories/<factory>/targets/intel-corei7-64-lmp-2377/meta/': 400 BAD REQUEST
        = Root metadata expire (2024-07-06T07:56:57Z) before the specified expiration time (2025-02-11T09:17:39Z)

        Getting production Target details; target: intel-corei7-64-lmp-2356, tag: master...
        Refreshing and downloading TUF metadata for Target intel-corei7-64-lmp-2356 to 2356/tuf...
        ERROR: Failed to download TUF metadata: HTTP error during POST 'https://api.foundries.io/ota/factories/<factory>/targets/intel-corei7-64-lmp-2356/meta/': 400 BAD REQUEST
        = Targets metadata expire (2025-01-28T16:38:23Z) before the specified expiration time (2025-02-06T09:27:35Z)

Use the ``--expires-days`` parameter of the ``fioctl targets offline-update`` command to set the desired expiration time of the bundle.
If the command fails because the root or targets metadata expires sooner than the date specified in the parameter,

> refresh the root or targets accordingly and then re-run the command.

Author: > The other option (which I prefer) is to provide a shorter expiration period.

IMO, this is not an option if a customer wants to set a specific expiration time for the bundle. Again, this is an instruction how to set "desired/specific" expiration time.

Also, it is kind of obvious that a shorter expiration time can help to download the bundle. But, it doesn't help to achieve the goal.
I'll let the customer support team to decide on this.

> refresh the root or targets accordingly and then re-run the command.

Reviewer 2: From my experience users at first try any random value for a parameter they want to try, and then try to analyse an error, and finally read the docs (as the last resort).
So it is hard to call it "desired".

Now, when they got to the docs, the documentation needs to tell the customer what _we want them to do_.
I personally do not wish them rotating the TUF root every time they try a random value as an input argument and see an error.
The best way is to explain them that the error itself is reasonable, and what options they have to resolve it.

> refresh the root or targets accordingly and then re-run the command.

Reviewer: I prefer guiding them through the behavior we want first (package should expire before the metadata expiration) then providing workarounds for increasing that period (rotating TUF root etc).

> refresh the root or targets accordingly and then re-run the command.

Author: The problem I have is that the root/targets can have just 1 day until it expires when the bundle command is fired. So, should we suggest them to decrease the expiration to a half-day?

>we want first (package should expire before the metadata expiration) 

Customer doesn't know the metadata expiration in advance.

Probably we need to say something like
"If cmd fails, then either:
a) set an earlier expiration time
or
b) refresh root and/or targets

depending on your needs?

> refresh the root or targets accordingly and then re-run the command.

Author: I have a problem with specifying something like this

>An end date specified by the ``--expire-days`` must be before than a minimum of root or targets metadata expiration

because we don't provide users with a way to find out this minimum conveniently.

> refresh the root or targets accordingly and then re-run the command.

Author: The same here

>The ``offline-updates`` command would fail if you try to provide a longer expiration period than allowed.

The first question that appears is "what period is allowed"?

> refresh the root or targets accordingly and then re-run the command.

Author: I just simply wrote 

>If the command fails because the root or targets metadata expires sooner than the date specified in the parameter, then
either decrease the parameter value or refresh the root/targets metadata accordingly and then re-run the command.

Maybe it makes sense to leave it as simple as it is.

> refresh the root or targets accordingly and then re-run the command.

Reviewer 2: > The problem I have is that the root/targets can have just 1 day until it expires when the bundle command is fired.

I think that by all means it is an exaggregation.

IIRC we notify the user well in advance about the TUF root going to expire, far more in advance than 1 day.
As for the targets, if they are going to expire, I don't think it is wise to cut an offline update based on that legacy.

So, under all logical circumstances, a user should have a long enough period in reserve.
But, if they are going to create a bundle "1 day before expiration" - there are 2 options:
- They don't know what they are doing; so showing an error is a better option than showing a workaround.
- They know what they are doing; so they will find a way, we don't need to encourage that wrong behavior.

Just IMHO.

> refresh the root or targets accordingly and then re-run the command.

Reviewer 2: > we don't provide users with a way to find out this minimum conveniently.

For this one, the `offline-update` command might show the maximum "allowed" expiration time when it fails because of exceeding it.
IMHO that would be the most convenient way to know it.

Other than that, `fioctl key tuf show-root` shows the TUF root expiration, while we also have a set of commands to show CI/wave/prod targets.json (for each tag) containing that data.

If a user does not know the expiration of their releases, and is not willing to check the above data, and does not like our error messages - sincerely, they've got a problem with their attitude, and we cannot help with that.

> refresh the root or targets accordingly and then re-run the command.

Reviewer 2: > I just simply wrote
> >  If the command fails because the root or targets metadata expires sooner than the date specified in the parameter, then
> >  either decrease the parameter value or refresh the root/targets metadata accordingly and then re-run the command.
> Maybe it makes sense to leave it as simple as it is.

I think it kind of makes sense.

I guess _vale_ blames on it, and I'd advice removing some wordiness... but that's @kprosise power, not mine.

> refresh the root or targets accordingly and then re-run the command.

Author: @vanmaegima @vkhoroz Please, see the latest version, and let me know if it is OK with you.

To refresh root role metadata use one of the ``fioctl keys tuf`` commands that update the root metadata,

> for example ``fioctl keys tuf updates rotate-offline-key --role=root``.

Author: I actually wanted the `tuf updates` sub-command, but I agree that the shorter version will serve better as an example.

> for example ``fioctl keys tuf updates rotate-offline-key --role=root``.

Author: > 2\. Suggestion: what about providing a link to where we document TUF keys?

I thought about it, I just was confused which of the docs to link to. Maybe "Offline Factory TUF Keys"?

> for example ``fioctl keys tuf updates rotate-offline-key --role=root``.

Author: Technically they can add another root key, not sure we should advise it in this case though.

> The expiration time is set to year since the moment when the root metadata is updated.

Author: What about a threshold update?

> The expiration time is set to year since the moment when the root metadata is updated.

Reviewer 2: Only the TUF root key addition or rotation extends the expiration time.
This is our security measure.

To refresh targets role metadata use one of the following depending on targets type, CI or wave/production.

* CI targets - Trigger a new CI build, it will create a new target and update CI targets role metadata expiration time to 1 year since the moment of creation.
* Wave or production targets - Create a new wave for the given target version.
  Use ``--expires-days`` or ``--expires-at`` parameters of the ``fioctl waves init`` command to set a desired expiration time.

Therefore, the `--expires-days` parameter in the primary knob to tune the bundle expiration time up to 1 year (the maximum validity period of root metadata).
Root and/or CI/wave/production targets refreshing serves as the secondary mechanism that should be applied

> if the desired expiration time occurs later than the root's and/or the targets' expiration, respectively.

Author: I'll let the customer support team to judge on it. IMO, the summary is useful. My IMO is based on my conversion with the support team on this matter.

Understanding the Math Behind the Offline Update Bundle Expiration Time
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The expiration time of the bundle is determined by the expiration times of the TUF metadata it encompasses.
Specifically, it equals the minimum value among the expiration times across all TUF roles' metadata.
Let's review the expiration times for each TUF role, including the default values set by FoundriesFactory,

> and explore how users can manage them.

Reviewer 2: A shorter text is easier to follow.

When I read the second sentence, since its middle I was questioning myself why is it there at all, as it brings no new information (as per my perception) compared to the first sentence.

* CI/Wave/Production *root* role metadata
    The expiration time is set to 1 year from the moment a new version of the root role metadata is generated.

>     A new version is generated whenever a user makes any changes to TUF keys, such as key rotation, deletion, addition, and so forth.

Author: > The expiration time is set to 1 year from the time of the latest root key addition or rotation using Fioctl.

What about key deletion? Which of the commands/actions that updates root metadata doesn't extend the expiration time? I suppose just one - "set-threshold".

>     A new version is generated whenever a user makes any changes to TUF keys, such as key rotation, deletion, addition, and so forth.

Reviewer 2: I gave a precise information.  An answer to all your follow-up questions is in it.

>     A new version is generated whenever a user makes any changes to TUF keys, such as key rotation, deletion, addition, and so forth.

Author: I wonder why key deletion doesn't refresh root metadata.

>     A new version is generated whenever a user makes any changes to TUF keys, such as key rotation, deletion, addition, and so forth.

Author: DOne.

* CI *timestamp*, *snapshot*, and *targets* roles metadata

>     The default expiration time is set to 1 year.
>     Users have the option to overwrite this default value using the factory config parameter :ref:`tuf.targets_expire_after <def-tuf-expiration>`.
>     Additionally, the FoundriesFactory automatically refreshes these metadata for one month before their expiration.

Reviewer 2: I think you captured it properly. I would just add a bit more precision to the statements.
```suggestion
    The default expiration time is set to 1 year since the last build.
    If there were no builds for more than a year, expiration is automatically extended by 1 month every month.
    You may overwrite this default value using the factory config parameter :ref:`tuf.targets_expire_after <def-tuf-expiration>`.
```

Links:
    - https://github.com/foundriesio/ci-scripts/blob/master/apps/publish.sh#L34
    - https://github.com/foundriesio/ota-lite/blob/main/ota_lite/tuf.py#L213

Author: > If there were no builds for more than a year, expiration is automatically extended by 1 month every month.

I am not sure this is quite correct. It is no "more than a year" rather a few days before the year period.

Also, not all builds produces targets and make a post/put request to the handler.

Reviewer 2: If not all builds produce targets, then we need to give an exact information which ones do it.

Currently, a user reading this is under impression that CI targets always have a 1 year expiration.
But,
1. It is not true, as sometimes they have 1 month expiration.
2. It is not clear what is a starting point for expiration period (is it always one year form now?).

I tried to answer these two questions which bothered me.
If you have a more precise information, you're welcome to share it in the documentation.

> I am not sure this is quite correct. It is no "more than a year" rather a few days before the year period.

I think it is justifiable to omit the "few days before the year period" part.
We never know how many days that "before" may happen, and technically it is there only for redundancy considerations.
A principal information here is that after (approximately) one year without targets being refreshed by a user, we start to auto-refresh them. But in that case the expiration time is different.

Author: I think I addressed this one.

* Wave/Production *timestamp* roles metadata
    The expiration time is set to 7 days.
    The `TUF specification`_ recommends setting a short expiration date and re-signing it frequently.
    This allows clients to quickly detect if they are being prevented from obtaining the most recent metadata ("indefinite freeze attacks").

>     The FoundriesFactory automatically refreshes the metadata for an additional 7 days just before expiration.

Reviewer 2: I think for the offline update you extend this period, right?
I believe it is worth mentioning that.

* Wave/Production *snapshot* and *targets* roles metadata
    The default expiration time is set to 1 year.
    A user can overwrite the default value using the ``--expires-days`` or ``--expires-at parameter`` of the ``fioctl wave init`` command.

The ``--expires-in-days`` parameter of the ``fioctl targets offline-update command`` instructs it to set
the expiration time of the TUF timestamp role metadata to the one specified by the user.
It's important to note that the user-specified expiration time is applied only to the bundle's copy

> of the metadata and does not affect the factory's metadata.

Reviewer 2: I see you doing it here.... but it really belongs to where the timestamp role is described.

Author: I just wanted to list facts - what foundriesfactory does by default in general, no use-case specifics.

And then after listing the general facts, state what is related to the `fioctl targets offline-update` command.

Reviewer 2: At first I added a comment to where you described "timestamp" to add this information. Because I had questions when reading that section.

But later I found the information I was missing in this paragraph. So, I removed my original comment, and am just asking to move this paragraph to where it answers my questions.

An information in question is this:
> The ``--expires-in-days`` parameter of the ``fioctl targets offline-update command`` instructs it to set
the expiration time of the TUF timestamp role metadata to the one specified by the user.

The question I had was "how the hell can I set my bundle expiration to 1 month if timestamp is always 7 days, and I cannot set a longer expiration than a minimum of all roles?".

Author: I strongly believe that this section except the last paragraph should be generic and describes the way we manage TUF meta in general. It should not contain any details about any specific details, in particular about offline update.
Moreover, I believe, that this info on TUF meta should be moved to some other spot where we provides users with info on TUF and this page just references it.

Author: Maybe this paragraph except the last sentence belongs here https://docs.foundries.io/latest/reference-manual/security/offline-keys.html#recommended-offline-tuf-keys-schema.

Reviewer 2: I'm not sure I get how the offline-updates specific use cases (about how the "overall expiration" is calculated) belongs to a section where we list our recommendations for TUF keys schema (topology).

This whole expiration topic only becomes a problem for the offline update due to its unique "bundle" approach. It is irrelevant for the most part in a regular (online update) use case.

Reviewer 2: > I strongly believe that this section except the last paragraph should be generic and describes the way we manage TUF meta in general.

Maybe, it is up to you.

I just think it is important to explicitly tell the user that their "bundle" expiration is _de facto_ calculated; not just how all the TUF timeouts play together. Then it suddenly turns out that for the most part the "expiration" is determined by the `expire-days` attribute. From that perspective the thing becomes simpler to understand and describe:
1. The "bundle" expiration equals a value specified in `expire-days`.
2. The `expire-days` has a hard limit which is a minimum between root and targets expiration.
3. So, we should probably also describe how the root and targets expiration works.
4. And the whole timestamp/snapshot expiration becomes irrelevant IIUC, as it has zero impact onto the "bundle" expiration.

It is just thoughts after me _walking the dog_ on it.
That said, there are many perspectives, I have no idea which one is the best point of view.

Author: > I'm not sure I get how the offline-updates specific use cases (about how the "overall expiration" is calculated) belongs to a section where we list our recommendations for TUF keys schema (topology).

Me neither. I couldn't find better spot.
I was looking for a doc which informs users about details of the TUF framework utilization/application in FoundriesFactory in general and about TUF role metadata specifically. I couldn't find it, so I just put as short as possible section into this offline related doc. My intention was just put here a reference to some other user doc which includes info about role expiration.

Author: > Maybe you want a new section out there, no?

I thought about it. The problem I have is that this section about metadata expiration looks to me as part of the bigger doc on the TUF application in FoundriesFactory. I don't want to start writing this bigger doc now, so I just put a section here. Once or if such bigger doc appears over the time then we can move the section from this spot to the new doc.

Author: > It is irrelevant for the most part in a regular (online update) use case.

I am not sure about it. One of the first questions that I would ask if I were a customer is "for how long are my targets valid?" and "when my tuf keys expire?" and "why timestamp is valid just for 7 days?", and so on.

Author: > Then it suddenly turns out that for the most part the "expiration" is determined by the `expire-days` attribute (and nothing else). From that perspective the thing becomes simpler to understand and describe:
> 
>     1. The "bundle" expiration equals a value specified in `expire-days`.
> 
>     2. The `expire-days` has a hard limit which is a minimum between root and targets expiration.
> 
>     3. So, we should probably also describe how the root and targets expiration works. (Or maybe not, and simply put links to somewhere).
> 
>     4. Then the whole timestamp/snapshot expiration becomes irrelevant IIUC, as it has zero impact onto the "bundle" expiration.
> 
> 
> It is just thoughts after me _walking the dog_ on it. That said, there are many perspectives, I have no idea which one is the best point of view.

That's almost exactly what I wrote in the "Controlling the Expiration Time of the Offline Update Bundle" section, just another wording. There is no #4 in that section btw.

My idea was and still is first of all to provide a hands-on instruction to deal with the bundle expiration - hence "Controlling the Expiration Time of the Offline Update Bundle" section. Basically it says, use `--expires-in-days` to set expiration time. If it fails then just decrease the value (lower your ambitions) (TBD as you and Vanessa suggested)  or refresh the root and/or targets meta. (Imperative part)

Secondarily, provide a user with a context info about TUF metadata expiration in a scope of FoundriesFactory. This is kind of bigger picture in case if a user would like to dig deeper. (story part).

If I were a customer I would seek for the "story" part first of all because I really do not like executing command without understanding of bigger picture :).

If you strongly believe that the story part is useless and @vanmaegima @caiotpereira thinks the same I am OK to remove it all.

Reviewer 2: @mike-sul I think this document is a good enough place for this info.
We can later revise this if we find a better place.

As for the story itself, I'm not against it... I was just trying to look at it from a different perspective, hoping that could help you write an even better story.
My [long comment](https://github.com/foundriesio/docs/pull/669#discussion_r1491475926) was in no way imperative... I thought I made it clear.

Author: @vkhoroz Please, the the latest version. Let's keep that "story" section in this doc.
 
However, I still think that we need a new doc which outlines how we utilize the TUF in our solution from a top level view. We do have several docs on the TUF related stuff, but they rather "nuts&bolts" (mechanics), not conceptual - describing the idea from top level perspective. If we do the new doc, then this doc should include references to the existing "nuts&bolts" documents.

2. Run one of :ref:`the post installation actions <Post Install and Run Actions>` depending on the ``aklite-offline install`` result:

    a. code 100: reboot device and invoke ``aklite-offline run [--config <config dir or file>]`` to finalize an ostree installation and start Apps if both ostree/rootfs and Apps are updated;
    b.  - (for LmP < v92) code 101: restart the Docker Engine (e.g. ``systemctl restart docker``) and invoke ``aklite-offline run [--config <config dir or file>]``  if just Apps are updated.

>         - (for LmP >= v92) code 10: just invoke ``aklite-offline run [--config <config dir or file>]``  tp start updated Apps.

Reviewer: typo: to start updated Apps.

Author: fixed. thx.

Offline Update Considerations
-----------------------------

* **Offline Update Bundle Packaging**

  The content provided by ``fioctl targets offline-update`` command should be packaged by you, and verified by the client service.

* **Offline Update Bundle Delivery**

  Related to the bullet above, Foundries.io™ cannot provide secure delivery of an update bundle since you should do the packaging and delivery.

* **Offline Update for Unregistered Devices**

>   When dealing with devices not registered in the FoundriesFactory® or :ref:`a custom registration server <ref-fully-detached>`,

Reviewer 3: ```suggestion
  When dealing with devices not registered in FoundriesFactory® or :ref:`a custom registration server <ref-fully-detached>`,
```

Author: Fixed.

Offline Update Considerations
-----------------------------

* **Offline Update Bundle Packaging**

  The content provided by ``fioctl targets offline-update`` command should be packaged by you, and verified by the client service.

* **Offline Update Bundle Delivery**

  Related to the bullet above, Foundries.io™ cannot provide secure delivery of an update bundle since you should do the packaging and delivery.

* **Offline Update for Unregistered Devices**

  When dealing with devices not registered in the FoundriesFactory®, several considerations arise:

  * Production Status: The distinction between production and non-production status of the device remains undetermined.

>   * Device Tag: The specific tag associated with the device is unavailable.

Author: It means "a tag it is not configured on the device?"

>   * Device Tag: The specific tag associated with the device is unavailable.

Reviewer 2: Got it, then what about this?
```suggestion
* Device Tag: The specific tag associated with the device is not configured.
```
It seems to be more precise to me, given your above clarification.

>   * Device Tag: The specific tag associated with the device is unavailable.

Author: makes sense

>   * Device Tag: The specific tag associated with the device is unavailable.

Author: done

>   As a consequence, during the initial offline update, users can install both production and non-production targets on unregistered devices.

Author: >When the user does everything properly, the unregistered device is flashed with a pre-loaded production TUF root role.

Not exactly, an unregistered device is flashed with both the CU and prod root role metadata.

>   As a consequence, during the initial offline update, users can install both production and non-production targets on unregistered devices.

Reviewer 2: Sorry, @mike-sul for my negligence of this topic, I'm trying to understand.

Can you be more specific here?
Do you mean that the "unregistered" device is flashed with something like `1.prod.root.json, 1.ci.root.json, 2.prod.root.json, 2.ci.root.json` in some folder?
How does the `offline-update` tool or `libaktualizr` know which of the pre-loaded TUF root roles need to be loaded?

Do we have this process documented somewhere?
If we do - it would be really nice to have that linked here.
Otherwise, did you consider documenting it?

>   As a consequence, during the initial offline update, users can install both production and non-production targets on unregistered devices.

Author: >Do you mean that the "unregistered" device is flashed with something like 1.prod.root.json, 1.ci.root.json, 2.prod.root.json, 2.ci.root.json in some folder?

Conceptually yes, technically it is stored in a bit different way:

```
root@raspberrypi4-64:/var/rootdirs/home/fio# tree -L 1 /usr/lib/sota/tuf
/usr/lib/sota/tuf
|-- ci
`-- prod
```

>   As a consequence, during the initial offline update, users can install both production and non-production targets on unregistered devices.

Author: >Do we have this process documented somewhere?
If we do - it would be really nice to have that linked here.
Otherwise, did you consider documenting it?

I don't think we need to document it. This is kind of internal implementation detail. At least, I don't recall that anyone (customers) was interested in such the details since the offline update was released.

  However, subsequent updates are constrained by the type of targets installed during the initial update.
  Therefore, the target type chosen during the first update dictates the supported targets' type for future updates.

  Additionally, due to the absence of a defined tag in the configuration of unregistered devices,

>   users can install targets associated with any tag.

Author: >My assumption is that you're talking about devices manufactured by flashing some image and generating a client certificate via fully detached mode (e.g. via lmp-dev-reg and factory-registration-ref).

No, this is a device which is not registered in any of the possible ways.

>   users can install targets associated with any tag.

Author: >So, in general, I think that your statement is actual for any device with unconfigured tag; and that has no correlation with the device being registered or not.

This is how it has worked for more than 2 years. There are customers that uses unregistered devices fully offline.

>   users can install targets associated with any tag.

Reviewer 2: @mike-sul thanks for the clarification.
My comment kind of suggested that I have no idea what do you mean by "unregistered" here.

I think it deserves defining somewhere what is an unregistered device (i.e. a device without `/var/sota/sota.toml`), so that there is no ambiguity. The word "unregistered" is overloaded here, so 3 different people may think of 3 different things.

If we have this term defined somewhere already - then we need to link it here.

>   users can install targets associated with any tag.

Reviewer 2: If I don't understand something - it does not mean it is not clear.
Whatever works for customer support and our end users, should be fine.

I'm just asserting that I understood it differently, out of my negligence. :-)

>   users can install targets associated with any tag.

Reviewer: We can introduce something like below on line 185 for more clarity:

  When dealing with unregistered devices, either before registration to FoundriesFactory® or for full offline devices, several considerations arise:

>   users can install targets associated with any tag.

Author: The proposed is not a definition of "unregistered device". Also, "full offline device" can be registered at the custom registration factory or even at the Foundries's backend during provisioning, and then operate in a fully offline mode. Such the device is actually registered.

>   users can install targets associated with any tag.

Reviewer 2: I think you latest suggestion kind of works, because one can deduct from it that `lmp-dev-reg` was not run, so most probably there is no config etc etc etc.

>   users can install targets associated with any tag.

Author: I think it is suffice. I really wouldn't like to touch the "configuration" subject, since this is a very vague term in our solution. And saying/writing "no config" can lead to misunderstanding (e.g. absence of `sota.toml` does not mean that a device is not configured because there are always some `*.toml` files in `/usr/sota/conf.d` and `/usr/sota/conf.d` by default (defined in yocto recipes in meta-lmp)).

>   users can install targets associated with any tag.

Author: updated.

    However, since the API supports both types of the update, a user may develop :ref:`a custom sota client <ug-custom-sota-client>` that does
    these two types of update in parallel by mistake.

Controlling the Expiration Time of the Offline Update Bundle
------------------------------------------------------------

The bundle obtained through the ``fioctl targets offline-update`` command comes with an expiration time.
If the expiration time of the bundle has passed, the offline update will fail.

> Use the ``--expires-in-days`` parameter of the ``fioctl targets offline-update`` command to set the desired expiration time of the bundle.

Reviewer 2: This is where I was expecting to see this explanation:
```
This parameter sets the expiration time for the bundle's copy of the TUF timestamp role metadata, and does not affect the factory's metadata.
```
(Copied from your text below).

As is, it is easy to lose this piece of information; and that's why I was asking my _silly_ question about the timeout role expiration.

If the command fails with the error below:

.. code-block:: bash

    Getting CI Target details; target: intel-corei7-64-lmp-2377, tag: master...
    Refreshing and downloading TUF metadata for Target intel-corei7-64-lmp-2377 to 2377/tuf...
    ERROR: Failed to download TUF metadata: HTTP error during POST 'https://api.foundries.io/ota/factories/<factory>/targets/intel-corei7-64-lmp-2377/meta/': 400 BAD REQUEST
    = Root metadata expire (2024-07-06T07:56:57Z) before the specified expiration time (2025-02-11T09:17:39Z)

    Getting production Target details; target: intel-corei7-64-lmp-2356, tag: master...
    Refreshing and downloading TUF metadata for Target intel-corei7-64-lmp-2356 to 2356/tuf...
    ERROR: Failed to download TUF metadata: HTTP error during POST 'https://api.foundries.io/ota/factories/<factory>/targets/intel-corei7-64-lmp-2356/meta/': 400 BAD REQUEST
    = Targets metadata expire (2025-01-28T16:38:23Z) before the specified expiration time (2025-02-06T09:27:35Z)

> , then it means the root or targets metadata expires sooner than the date specified in the parameter.

Reviewer 3: Good catch

> , then it means the root or targets metadata expires sooner than the date specified in the parameter.

Author: @kprosise Please, comment on this.

> , then it means the root or targets metadata expires sooner than the date specified in the parameter.

Author: what's the fix, just remove comma?

> , then it means the root or targets metadata expires sooner than the date specified in the parameter.

Author: updated via a forced push.

To fix the issue, either decrease the parameter value or refresh the root/targets metadata accordingly and then re-run the command.

To refresh root metadata you should :ref:`rotate TUF root role key <ref-offline-tuf-root-key-rotation>`.
The expiration time is set to one year since the moment of the latest root key rotation.

To refresh targets role metadata use one of the following depending on targets type, CI or wave/production.

* CI targets — Trigger a new CI build, it will create a new target and update CI targets role metadata expiration time to 1 year since the moment of creation.
* Wave or production targets — Create a new wave for the given target version.

>   Use ``--expires-days`` or ``--expires-at`` parameters of the ``fioctl waves init`` command to set a desired expiration time.

Author: won't harm, I guess.

>   Use ``--expires-days`` or ``--expires-at`` parameters of the ``fioctl waves init`` command to set a desired expiration time.

Author: updated via forced push.

Specifically, it equals the minimum value among the expiration times across all TUF roles' metadata.

* CI/Wave/Production *root* role metadata
    The expiration time is set to 1 year from the moment when of the latest TUF root key was added or rotated.
    The other commands that modify the TUF root metadata do not extend its expiration.
    It is possible to set the TUF root expiration time to any value through the API.

* CI *timestamp*, *snapshot*, and *targets* roles metadata
    The default expiration time is set to 1 year since the last successful CI build.
    If there are no builds for a year, the expiration is automatically extended by one month every month.
    A user can overwrite the default value using the factory config parameter :ref:`tuf.targets_expire_after <def-tuf-expiration>`.

* Wave/Production *timestamp* roles metadata
    The expiration time is set to 7 days.
    The `TUF specification`_ recommends setting a short expiration date for the TUF timestamp metadata and re-signing it frequently.
    This allows clients to quickly detect if they are being prevented from obtaining the most recent metadata ("indefinite freeze attacks").

>     The FoundriesFactory automatically refreshes the metadata for an additional 7 days just before expiration.

Author: >I'm under the impression that the production bundle will expire in 7 days, no matter what the user does. But we both know this is not the case.

That's true. This section is not about a bundle's TUF meta, this is about the factory's TUF meta.

The section called `Controlling the Expiration Time of the Offline Update Bundle` is for the bundle meta, and it has this statement `Effectively, this parameter sets the expiration time for the bundle's copy of the TUF timestamp role metadata, and does not affect the factory's metadata.`

>     The FoundriesFactory automatically refreshes the metadata for an additional 7 days just before expiration.

Author: Just to be clear, this section is not about the bundle's meta, it is about factory's meta (that's why earlier I proposed moved it somewhere else).

>     The FoundriesFactory automatically refreshes the metadata for an additional 7 days just before expiration.

Reviewer 2: Thanks, I found that sentence... it was apparently hiding from me.
With it, this comment is no more relevant.

  * Device Tag: The specific tag associated with the device is unavailable.

  As a consequence, during the initial offline update, users can install both production and non-production targets on unregistered devices.
  However, subsequent updates are constrained by the type of targets installed during the initial update.
  Therefore, the target type chosen during the first update dictates the supported targets' type for future updates.

  Additionally, due to the absence of a defined tag in the configuration of unregistered devices,
  users can install targets associated with any tag.
  This issue can be addressed by incorporating a configuration snippet (a ``*.toml`` file) into either ``/usr/lib/sota/conf.d`` or ``/etc/sota/conf.d``.
  We recommend implementing this solution through a new recipe in the factory's ``meta-subscriber-overrides.git`` repository.
  The snippet should contain the following content:

.. prompt:: text

    [pacman]
    tags = "<tag>"

* **Online/Offline Mixed Updates (aka hybrid mode)**

  There are a few points to take into account by the custom client application:

  * Offline or Online downgrade fails by default.

>     Therefore, if the latest target is installed on a device through an online update, then an offline update for the outdated bundle fails.

Reviewer: @mike-sul this is more for my information then an actual suggestion to the docs: we have customers relying on the downgrade behavior, in the past we discussed that this could be achieved by a --force operation, is this still true?

Author: Yes, the `--force` option is still there, that's why I stated "by default".

Author: @vanmaegima Please, see the latest version, I added the `unless...` statement.

  We recommend implementing this solution through a new recipe in the factory's ``meta-subscriber-overrides.git`` repository.
  The snippet should contain the following content:

.. prompt:: text

    [pacman]
    tags = "<tag>"

* **Online/Offline Mixed Updates (aka hybrid mode)**

  There are a few points to take into account by the custom client application:

  * Offline or Online downgrade fails by default.
    Therefore, if the latest target is installed on a device through an online update,
    then an offline update for the outdated bundle fails, unless a user explicitly specifies the ``--force`` parameter.
  * Offline Update fails if its input TUF metadata are outdated.
    For example, it fails if an online update upgrades the device's TUF root/timestamp/snapshot/targets metadata
    to version N while the bundle contains version N-1 of the metadata.
  * Running offline and online update simultaneously leads to undefined behaviour.
    It is impossible to run two or more update agents of the same or different types simultaneously,
    for example ``aktualizr-lite daemon`` and ``aklite-offline``.

>     However, since the API supports both types of the update, a user may develop a custom sota client that does

Reviewer: Can we link the custom sota client page here?

  * Device Tag: The specific tag associated with the device is unavailable.

  As a consequence, during the initial offline update, users can install both production and non-production targets on unregistered devices.
  However, subsequent updates are constrained by the type of targets installed during the initial update.
  Therefore, the target type chosen during the first update dictates the supported targets' type for future updates.

  Additionally, due to the absence of a defined tag in the configuration of unregistered devices, users can install targets associated with any tag.
  This issue can be addressed by incorporating a configuration snippet (a *.toml file) into  either ``/usr/lib/sota/conf.d`` or ``/etc/sota/conf.d``.
  We recommend implementing this solution through a new recipe in the factory's ``meta-subscriber-overrides.git`` repository.
  The snippet should contain the following content:

.. prompt:: text

    [pacman]
    tags = "<tag>"

* **Online/Offline Mixed Updates (aka hybrid mode)**

  There are a few points to take into account by the custom client application:

>   * Offline or Online downgrade fails by default. A device can be updated online with the latest target, so an offline update for the outdated bundle fails.

Reviewer 2: It is hard to understand how the second sentence adds to the first one.
I mean:
- The first sentence is imperative: it basically tells that some action is not allowed.
- The second sentence is descriptive: it tells the story of a specific use case.

A grammar sugar like `For example,` might help to tie them together.

```suggestion
* Offline or Online downgrade fails by default.
  For example, if the latest target was installed on a device using online update, an offline update for the outdated bundle fails.
```

Author: I think, "therefore" would serve better since this is the "considerations" section.

Reviewer 2: Yes, I think using `Therefore` instead of `For example` may be justifiable.
The most important part here is to use a conditional sentence like `if something happens, a consequence comes`.

>   * Offline Update fails if its input TUF metadata are outdated. For example, an online update updated device's TUF root/timestamp/snapshot/targets metadata to version N while an offline content has version N-1 of the metadata.

Reviewer 2: Similar to the above, but this one is better.

I'd simply add this:
```suggestion
  * Offline Update fails if its input TUF metadata are outdated. For example, it fails if an online update updated device's TUF root/timestamp/snapshot/targets metadata to version N while an offline content has version N-1 of the metadata.
```

And then apply semantic newlines, and maybe reword to make it more concise.

Author: I disagree with this one.

Reviewer 2: Without the `it fails if` part, I do not understand what that second sentence is about.
I kind of deducted that missing part from the first sentence.

If this is not what you wanted to say, then try adding something or paraphrasing it to make your motives clear.

Author: I think I addressed it in the latest version (forced push).

* **Online/Offline Mixed Updates (aka hybrid mode)**

   * Offline Update can downgrade version: A client around the offline updater should check it out and decide whether to allow a downgrade or not.
  There are a few points to take into account by the custom client application:

   * Offline Update does content-based shortlisting: Only the Apps included in a source directory are installed.
  * Offline or Online downgrade fails by default. A device can be updated online with the latest target, so an offline update for the outdated bundle fails.
  * Offline Update fails if its input TUF metadata are outdated. For example, an online update updated device's TUF root/timestamp/snapshot/targets metadata to version N while an offline content has version N-1 of the metadata.
  * Running offline and online update simultaneously leads to undefined behaviour.
    It is impossible to run two or more update agents of the same or different types simultaneously. For example ``aktualizr-lite daemon`` and ``aklite-offline``.
    However, since the API supports both types of the update, a user may develop a custom sota client that does these two types of update in parallel by mistake.

>    * Offline Update fails if its input TUF metadata are outdated, e.g. an online update updated TUF root meta to version N while an offline content has version N-1 of root meta.

Reviewer 2: Can you use semantic newlines everywhere?
